File: SEGUNDO/C2/PDOO/backup/DeepSpace_ruby/pkg/DeepSpace_ruby-0.0.1/lib/game_universe.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.
require_relative 'SpaceStation'
require_relative 'GameStateController'
require_relative 'Dice'
require_relative 'EnemyStarShip'
require_relative 'GameState'
require_relative 'GameUniverseToUI'
require_relative 'CombatResult'
require_relative 'CardDealer'
require_relative 'ShotResult'
require_relative 'GameCharacter'

module Deepspace
class GameUniverse
  @@WIN=10
  
  def initialize()
    @currentStationIndex=0
    @turns=0
    @currentStation=nil
    @spaceStations=Array.new
    @dice=Dice.new
    @gameState=GameStateController.new
    @currentEnemy=nil
  end
  
  def combatGo(station,enemy)
    ch=@dice.firstShot
      
      if(ch == GameCharacter::ENEMYSTARSHIP)
        fire = @currentEnemy.fire
        result = @currentStation.receiveShot(fire)
        
        if(result == ShotResult::RESIST)
          fire = @currentStation.fire
          
          result = @currentEnemy.receiveShot(fire)
          
          enemyWins =(result == ShotResult::RESIST)
          
        else
          enemyWins=true
        end
      
      else
        fire = @currentStation.fire
        result = @currentEnemy.receiveShot(fire)
        
        enemyWins=(result==ShotResult::RESIST)
      end
      
      if(enemyWins)
        s=@currentStation.getSpeed
        
        moves = @dice.spaceStationMoves(s)
        
        if(!moves)
          damage = @currentEnemy.damage
          
          @currentStation.setPendingDamage(damage)
          
          combatResult=CombatResult::ENEMYWINS
          
        else
          @currentStation.move
          combatResult=CombatResult::STATIONESCAPES
        end
        
      else
        aLoot = @currentEnemy.loot
        
        @currentStation.setLoot(aLoot)
        
        combatResult=CombatResult::STATIONWINS
      end
      
      @gameState.next(@turns,@spaceStations.length)
      
      return combatResult
  end 
  
  def combat
    state=@gameState.state
    
    if(state == GameState::BEFORECOMBAT || state == GameState::INIT)
      return self.combatGo(@currentStation,@currentEnemy)
      
    else
      return CombatResult::NOCOMBAT
    end 
    
  end
  
  def discardHangar
    if(@gameState.state == GameState::INIT || @gameState.state == GameState::AFTERCOMBAT)
      @currentStation.discardHangar
    end
  end
  
  def discardShieldBooster(i)
    if(@gameState.state == GameState::INIT || @gameState.state == GameState::AFTERCOMBAT)
      @currentStation.discardShieldBooster(i)
    end
  end
  
  def discardShieldBoosterInHangar(i)
    if(@gameState.state == GameState::INIT || @gameState.state == GameState::AFTERCOMBAT)
      @currentStation.discardShieldBoosterInHangar(i)
    end
  end
  
  def discardWeapon(i)
    if(@gameState.state == GameState::INIT || @gameState.state == GameState::AFTERCOMBAT)
      @currentStation.discardWeapon(i)
    end
  end
  
  def discardWeaponInHangar(i)
    if(@gameState.state == GameState::INIT || @gameState.state == GameState::AFTERCOMBAT)
      @currentStation.discardWeaponInHangar(i)
    end
  end
  
  def state
    if(@gameState.state == GameState::INIT)
        return GameState::INIT
      else if(@gameState.state == GameState::CANNOTPLAY)
          return GameState::CANNOTPLAY
      else if(@gameState.state == GameState::BEFORECOMBAT)
          return GameState::BEFORECOMBAT
      else
        return GameState::AFTERCOMBAT
      end
    end
    end
  end
  
  def getUIversion
    return GameUniverseToUI.new(@currentStation,@currentEnemy)
  end
  
  def haveAWinner
    if(@currentStation.nMedals == @@WIN)
      true
      
    else
      false
    end    
  end
  
  def init(names)
    state=@gameState.state
    
    if(state == GameState::CANNOTPLAY)
      @spaceStations=Array.new
      
      dealer=CardDealer.instance
      
      for i in 0..(names.length-1)
        supplies = dealer.nextSuppliesPackage
        station = SpaceStation.new(names.at(i), supplies)
        
        @spaceStations.push(station)
        
        nh = @dice.initWithNHangars
        nw = @dice.initWithNWeapons
        ns = @dice.initWithNShields
        
        lo = Loot.new(0,nw,ns,nh,0)
        
        station.setLoot(lo)
      end
      
      @currentStationIndex = @dice.whoStarts(names.length)
      
      @currentStation = @spaceStations.at(@currentStationIndex)
      
      @currentEnemy = dealer.nextEnemy
      
      @gameState.next(@turns, @spaceStations.length)
    end
  end
  
  def mountShieldBooster(i)
    if(@gameState.state == GameState::INIT || @gameState.state == GameState::AFTERCOMBAT)
      @currentStation.mountShieldBooster(i)
    end
  end
  
  def mountWeapon(i)
    if(@gameState.state == GameState::INIT || @gameState.state == GameState::AFTERCOMBAT)
      @currentStation.mountWeapon(i)
    end
  end
  
  def nextTurn
    state=@gameState.state
    
    if(state==GameState::AFTERCOMBAT)
      stationState=@currentStation.validState
      
      if(stationState)
        @currentStationIndex=(@currentStationIndex + 1)%@spaceStations.length
        @turns += 1
        
        @currentStation=@spaceStations.at(@currentStationIndex)
        
        @currentStation.cleanUpMountedItems
        
        dealer=CardDealer.instance
        
        @currentEnemy=dealer.nextEnemy
        
        @gameState.next(@turns,@spaceStations.length)
        
        return true
      end
      
      return false
    end
    
    return false
  end
  
  def to_s
    salida="CurrentStationIndex: #{@currentStationIndex}\nTurnos: #{@turns}\nCurrentStation: #{@currentStation.to_s}\nDice: #{@dice.to_s}\n"
    
    if(@gameState.state==GameState::INIT)
      salida=salida +"Estado: iniciando\n" 
    elsif(@gameState.state==GameState::AFTERCOMBAT)
      salida=salida + "Estado: aftercombat\n"
    elsif(@gameState.state==GameState::BEFORECOMBAT)
      salida=salida + "Estado: beforecombat\n"
    else
      salida=salida + "Estado: cannotplay\n"
    end
    
    salida= salida + "Enemigo Actual: #{@currentEnemy.to_s}\n"
  end
  
end
end

File: SEGUNDO/C2/PDOO/backup/Deep_space/src/deepspace/GameUniverse.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package deepspace;

import java.util.ArrayList;
/**
 *
 * @author daniel
 */
public class GameUniverse {
    static private int WIN=10;
    private int currentStationIndex;
    private int turns;
    private GameStateController gameState;
    private SpaceStation currentStation;
    private ArrayList<SpaceStation> spaceStations;
    private Dice dice;
    private EnemyStarShip currentEnemy;
    
    public GameUniverse(){
        this.gameState = new GameStateController();
        this.turns = 0;
        this.dice = new Dice();
    }
    
    CombatResult combat(SpaceStation station, EnemyStarShip enemy){
        GameCharacter ch = dice.firstShot();
        boolean enemyWins;
        CombatResult combatResult;
        
        if(ch == GameCharacter.ENEMYSTARSHIP){
            float fire = enemy.fire();
            
            ShotResult result = station.receiveShot(fire);
            
            if(result == ShotResult.RESIST){
                fire = station.fire();
                
                result = enemy.receiveShot(fire);
                
                
                enemyWins = (result == ShotResult.RESIST);
            }
            
            else{
                enemyWins = true;
            }   
        }
        
        else{
            float fire = station.fire();
            ShotResult result = enemy.receiveShot(fire);
            enemyWins = (result == ShotResult.RESIST);
        }
        
        if(enemyWins){
            float s = station.getSpeed();
            
            boolean moves = dice.spaceStationMoves(s);
            
            if(!moves){
                Damage damage = enemy.getDamage();
                station.setPendingDamage(damage);
                combatResult = CombatResult.ENEMYWINS;
            }
            
            else{
                station.move();
                
                combatResult = CombatResult.STATIONESCAPES;
            }
            
        }
        
        else{
            Loot aLoot = enemy.getLoot();
            
            station.setLoot(aLoot);
            
            combatResult = CombatResult.STATIONWINS;
        }
        
        gameState.next(turns, spaceStations.size());
        
        return combatResult;
    }
    
    public CombatResult combat(){
        GameState state=gameState.getState();
        
        if(state == GameState.BEFORECOMBAT || state == GameState.INIT){
            return this.combat(currentStation, currentEnemy);
        }
        
        else
            return CombatResult.NOCOMBAT;
    }
    
    public void discardHangar(){
        if(this.gameState.getState()==GameState.INIT || this.gameState.getState()==GameState.AFTERCOMBAT)
            this.currentStation.discardHangar();
    }
    
    public void discardShieldBooster(int i){
        if(this.gameState.getState()==GameState.INIT || this.gameState.getState()==GameState.AFTERCOMBAT)
            this.currentStation.discardShieldBooster(i);
    }
    
    public void discardShieldBoosterInHangar(int i){
        if(this.gameState.getState()==GameState.INIT || this.gameState.getState()==GameState.AFTERCOMBAT)
            this.currentStation.discardShieldBoosterInHangar(i);
    }
    
    public void discardWeapon(int i){
        if(this.gameState.getState()==GameState.INIT || this.gameState.getState()==GameState.AFTERCOMBAT)
            this.currentStation.discardWeapon(i);
    }
    
    public void discardWeaponInHangar(int i){
        if(this.gameState.getState()==GameState.INIT || this.gameState.getState()==GameState.AFTERCOMBAT)
            this.currentStation.discardWeaponInHangar(i);      
    }
    
    public GameState getState(){
        return this.gameState.getState();
    }
    
    public GameUniverseToUI getUIversion(){
        return new GameUniverseToUI(this.currentStation, this.currentEnemy);
    }
    
    public boolean haveAWinner(){
        if(this.currentStation.getNMedals() == WIN)
            return true;
        
        else
            return false;
    }
    
    public void init(ArrayList<String> names){
        GameState state = gameState.getState();
        
        if(state == GameState.CANNOTPLAY){
            spaceStations = new ArrayList<SpaceStation>();
            
            CardDealer dealer = CardDealer.getInstance();
            
            for(int i=0; i < names.size(); i++){
                SuppliesPackage supplies = dealer.nextSuppliesPackage();
                SpaceStation station = new SpaceStation(names.get(i), supplies);
                
                spaceStations.add(station);
                
                int nh = dice.initWithNHangar();
                int nw = dice.initWithWeapons();
                int ns = dice.initWithNShields();
                
                Loot lo = new Loot(0,nw,ns,nh,0);
                
                station.setLoot(lo);
            }
            
            currentStationIndex = dice.whoStarts(names.size());
            
            currentStation = spaceStations.get(currentStationIndex);
            
            currentEnemy = dealer.nextEnemy();
            
            gameState.next(turns, spaceStations.size());
        }
        
    }
    
    public void mountShieldBooster(int i){
        if(this.gameState.getState()==GameState.INIT || this.gameState.getState()==GameState.AFTERCOMBAT)
            this.currentStation.mountShieldBooster(i);
    }
    
    public void mountWeapon(int i){
        if(this.gameState.getState()==GameState.INIT || this.gameState.getState()==GameState.AFTERCOMBAT)
            this.currentStation.mountWeapon(i);
    }
    
    public boolean nextTurn(){
        boolean salida;
        GameState state = gameState.getState();
        
        if(state == GameState.AFTERCOMBAT){
            boolean stationState = currentStation.validState();
            
            if(stationState){
                currentStationIndex=(currentStationIndex + 1)%spaceStations.size();
                turns++;
                
                currentStation = spaceStations.get(currentStationIndex);
                currentStation.cleanUpMountedItems();
                
                CardDealer dealer = CardDealer.getInstance();
                
                currentEnemy = dealer.nextEnemy();
                
                gameState.next(turns, spaceStations.size());
                
                return true;
            }
            
            return false;
        }
        
        return false;
    }
    
    @Override
    public String toString(){
        String salida= "Medallas necesarias para ganar: " + WIN;
        salida = salida + "\nIndice Estación Actual: " + currentStationIndex + "\nTurnos: " + turns;
        
        if(gameState.getState() == GameState.AFTERCOMBAT)
            salida = salida + "\nEstado: AFTERCOMBAT";
        
        else if(gameState.getState() == GameState.BEFORECOMBAT)
            salida = salida + "\nEstado: BEFORECOMBAT";
        
        else if(gameState.getState() == GameState.CANNOTPLAY)
            salida = salida + "\nEstado: CANNOTPLAY";
        
        else
            salida = salida + "\nEstado: INIT";
        
        salida = salida + "\nEstación actual: " + currentStation.toString() + "\nLista de Estaciones Espaciales: \n";
        
        if(spaceStations != null){
            for(int i=0; i < spaceStations.size(); i++)
                salida = salida + spaceStations.get(i).toString();
        }
        
        salida = salida + "\nDice: " + dice.toString();
        salida = salida + "Nave Enemiga Actual: " + currentEnemy + "\n";
        
        return salida;        
    }
    
}
